File: pr003.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
#

'''
Problem 3
The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?
'''

INPUT_NUM = 600851475143

def has_factor(num, lst):
  '''check where lst has factor of num'''
  for i in lst:
    if num % i == 0:
      return True
  return False

def prime():
  '''generator for prime number'''
  primes = []
  num = 1  # we start at '2' (not include '1')
  while True:
    num += 1
    if has_factor(num, primes):
      continue
    primes.append(num)
    yield num

def prime_factor(num):
  '''return list of prime factors'''
  factors = []
  while num > 1:
    for p in prime():
      if num % p == 0:
        factors.append(p)
        num = num / p
        break
  return factors

def main(num):
  '''main function'''
  print max(prime_factor(num))

if __name__ == '__main__':
  main(INPUT_NUM)

File: pr003.hs
{-
Problem 3
The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?
-}

module Main where

has_factor :: Int -> [Int] -> Bool
has_factor _ [] = False
has_factor n (f:fs) | n `mod` f == 0 = True
                    | otherwise = has_factor n fs

next_prime :: Int -> Int
next_prime n = if has_factor (n + 1) [2..n]
                   then next_prime (n + 1)
                   else (n + 1)

pf :: Int -> [Int] -> Int -> [Int]
pf 1 fs _ = fs
pf n fs n2 | n `mod` n2 == 0 = pf (n `div` n2) (n2:fs) n2
           | otherwise = pf n fs (next_prime n2)

prime_factors :: Int -> [Int]
prime_factors n = pf n [] (next_prime 1)

main :: IO ()
main = print $ maximum $ prime_factors 600851475143

